tools/c7n_azure/c7n_azure/function_package.py

```diff
@@ -149,34 +154,43 @@ def _add_cffi_module(self):
     def _update_perms_package(self):
         os.chmod(self.pkg.path, 0o0644)
 
-    def build(self):
+    def build(self, policy, entry_point=None, extra_modules=None):
         # Get dependencies for azure entry point
-        modules, so_files = FunctionPackage._get_dependencies('entry.py')
+        entry_point = entry_point or \
+            os.path.join(os.path.dirname(os.path.realpath(__file__)), 'entry.py')
+        modules, so_files = FunctionPackage._get_dependencies(entry_point)
 
         # add all loaded modules
-        modules.remove('azure')
+        modules.discard('azure')
         modules = modules.union({'c7n', 'c7n_azure', 'pkg_resources'})
+        if extra_modules:
+            modules = modules.union(extra_modules)
+
         self.pkg.add_modules(None, *modules)
 
         # adding azure manually
         # we need to ignore the __init__.py of the azure namespace for packaging
         # https://www.python.org/dev/peps/pep-0420/
         self.pkg.add_modules(lambda f: f == 'azure/__init__.py', 'azure')
 
-        # add Functions HttpTrigger
-        self._add_functions_required_files()
+        # add config and policy
+        self._add_functions_required_files(policy)
 
         # generate and add auth
         s = local_session(Session)
-        self.pkg.add_contents(dest=self.policy['name'] + '/auth.json', contents=s.get_auth_string())
+        self.pkg.add_contents(dest=self.name + '/auth.json', contents=s.get_auth_string())
 
         # cffi module needs special handling
         self._add_cffi_module()
 
-        self.pkg.close()
-
-        # update perms of the package
-        self._update_perms_package()
+    def wait_for_status(self, app_name, retries=3, delay=30):
```

Comment:
Wow does it take 30 seconds for the Function App? (Seems super long)

Why don't we just move this "retry/status check logic" into the publish call and hide the fact you have to wait for status before attempting to publish.